internal/sync2/cycle.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package sync2
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// Cycle implements a controllable recurring event.
+//
+// Cycle control methods don't have any effect after the cycle has completed.
+type Cycle struct {
+	interval time.Duration
+
+	ticker  *time.Ticker
+	control chan interface{}
+	quit    chan struct{}
+
+	init sync.Once
+}
+
+type (
+	// cycle control messages
+	cyclePause    struct{}
+	cycleContinue struct{}
+	cycleTrigger  struct {
+		done chan struct{}
+	}
+)
+
+// NewCycle creates a new cycle with the specified interval.
+func NewCycle(interval time.Duration) *Cycle {
+	cycle := &Cycle{}
+	cycle.SetInterval(interval)
+	return cycle
+}
+
+// SetInterval allows to change the interval before starting.
+func (cycle *Cycle) SetInterval(interval time.Duration) {
+	cycle.interval = interval
+}
+
+func (cycle *Cycle) initialize() {
+	cycle.init.Do(func() {
+		cycle.quit = make(chan struct{})
+		cycle.control = make(chan interface{})
+	})
+}
+
+// Start runs the specified function with an errgroup
+func (cycle *Cycle) Start(ctx context.Context, group *errgroup.Group, fn func(ctx context.Context) error) {
+	group.Go(func() error {
+		return cycle.Run(ctx, fn)
+	})
+}
+
+// Run runs the specified function.
+func (cycle *Cycle) Run(ctx context.Context, fn func(ctx context.Context) error) error {
+	cycle.initialize()
+	defer close(cycle.quit)
+
+	currentInterval := cycle.interval
+	cycle.ticker = time.NewTicker(currentInterval)
+	if err := fn(ctx); err != nil {
+		return err
+	}
+	for {
+		select {
+
+		case message := <-cycle.control:
+			// handle control messages
+
+			switch message := message.(type) {
+			case nil:
+				return nil
+
+			case time.Duration:
+				currentInterval = message
+				cycle.ticker.Stop()
+				cycle.ticker = time.NewTicker(currentInterval)
+
+			case cyclePause:
+				cycle.ticker.Stop()
+				// ensure we don't have ticks left
+				select {
+				case <-cycle.ticker.C:
+				default:
+				}
+
+			case cycleContinue:
+				cycle.ticker.Stop()
+				cycle.ticker = time.NewTicker(currentInterval)
+
+			case cycleTrigger:
+				// trigger the function
+				if err := fn(ctx); err != nil {
+					return err
+				}
+				if message.done != nil {
+					message.done <- struct{}{}
+				}
+			}
+
+		case <-ctx.Done():
+			// handle control messages
+			return ctx.Err()
+
+		case <-cycle.ticker.C:
+			// trigger the function
+			if err := fn(ctx); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// sendControl sends a control message
+func (cycle *Cycle) sendControl(message interface{}) {
+	cycle.initialize()
+	select {
```

Comment:
so this either sends a message or quits ?

Reply:
This relies on closing behavior on channels. `<-quit` blocks until it's closed.